### PR TITLE
Feature/3 some performance optimalisations

### PR DESCRIPTION
**Pull request by zsinnema**

### CODE REVIEW

**Comment by zsinnema** [comment]
please enable chrome://flags/#enable-gpu-rasterization

File: src/model/chainModel.js [hunk 1]
```diff
@@ -212,31 +220,37 @@ export class Chain extends GroupTemplate {
...
-                };
-                sources[i] = sources[i] || targets[i] || newNode;
+                });
+
+            for (let i = 0; i < N; i++){
+                sources[i] = sources[i] || ((i > 0) && targets[i-1]) || getNewNode(i);
                 mergeGenResource(chain.group, parentGroup, sources[i], $Field.nodes);
             }
-            targets[targets.length - 1] = targets[targets.length - 1] || chain.leaf;
+            for (let i = 1; i < N; i++){
```
Comment (ddelpiano) [inline src/model/chainModel.js:246]: why don't merge these 2 loops into a single one?

File: src/model/chainModel.js [hunk 2]
```diff
@@ -319,9 +333,16 @@ export class Chain extends GroupTemplate {
                 return;
             }
 
+            let sameAsPrev = i > 0 && chain.housingLyphs[i] === chain.housingLyphs[i-1];
+            let sameAsNext = i < N - 1 && chain.housingLyphs[i] === chain.housingLyphs[i+1];
+            let sourceInternal = (i === 0);
+            let targetInternal = (i === N - 1) || sameAsNext && !sameAsPrev;
```
Comment (ddelpiano) [inline src/model/chainModel.js:339]: these 4 variables to const?

File: src/model/graphModel.js [hunk 3]
```diff
@@ -192,8 +192,10 @@ export class Graph extends Group{
         res.entitiesByID = entitiesByID;
 
         if (!res.generated) {
-            res.createAxesForInternalLyphs(modelClasses, entitiesByID, namespace);
-            res.createAxesForAllLyphs(modelClasses, entitiesByID, namespace);
+            let noAxisLyphsInternal = (res.lyphs||[]).filter(lyph => lyph.internalIn && !lyph.axis && !lyph.isTemplate);
+            res.createAxes(noAxisLyphsInternal, modelClasses, entitiesByID, namespace);
+            let noAxisLyphs = (res.lyphs||[]).filter(lyph => lyph::isObject() && !lyph.conveys && !lyph.layerIn && !lyph.isTemplate);
```
Comment (ddelpiano) [inline src/model/graphModel.js:197]: noAxisLyphsInternal and noAxisLyphs to const?

File: src/model/graphModel.js [hunk 4]
```diff
@@ -209,6 +211,16 @@ export class Graph extends Group{
         (res.links||[]).forEach(link => {
             if (link instanceof modelClasses.Link){
                 link.validateProcess();
+                if (!link.source.sourceOf){
+                    logger.error($LogMsg.NODE_NO_LINK_REF, link);
+                    console.log(link);
```
Comment (ddelpiano) [inline src/model/graphModel.js]: since the logger is already used, do we need the console.log as well?

File: src/model/graphModel.js [hunk 5]
```diff
@@ -209,6 +211,16 @@ export class Graph extends Group{
...
+                    return;
+                }
+                if (!link.target.targetOf){
+                    logger.error($LogMsg.NODE_NO_LINK_REF, link);
+                    console.log(link);
```
Comment (ddelpiano) [inline src/model/graphModel.js]: same here

**Review by ddelpiano — CHANGES_REQUESTED** [review]

**Pull request merged**